### ISSUE TITLE
unique dctemp paths when recursing archives

### ISSUE THREAD

**Issue by mikjo** [dependency-check/DependencyCheck#530]
Previously, when scanning zip files with archives included, the `filePath` in `dependencies` and `relatedDependences` in the dependency-check-report.xml file would list a path through the zip file:

```
<filePath>/path/to/container.zip/directory/below.jar</filePath>
```

With 1.4.0, we instead see the `filePath` with a transient temporary path (e.g. `/tmp/dctemp/check1868579570325300612tmp/1/directory/below.jar`). This is causing us two problems:
1. The Jenkins plugin sees each of those paths as unique, and so is reporting massive churn in "fixed" and "new" vulnerabilities that hide the signal of real changes in noise of unique temporary paths.
2. More importantly, it drops from the report the actual attribution required to fix the problems. We simply do not know from the report where the vulnerability is present.

This makes me think that there is a bug or design error in the `actualFilePath` and `packagePath` handling changes; a temporary directory (by design changing name for each run) in the report can't really be useful information for the user.

For some of our use cases, we are working in temporary directories and can just unpack archives before invoking OWASP DC; that has worked so far. However, in others we need to scan read-only (to us) directories where this workaround isn't available.

Annoyingly, while we've seen this many times in production, our initial simple attempts to create a minimal test case have failed. (We'll work on that, but didn't want to wait for success before reporting it in case others are running into the same problem.)

This looks related to but not the same as #476

**Comment by mikjo** [comment]
@wschiang and I have looked at this together

**Comment by stevespringett** [comment]
Regardless of how the Dependency-Check results are generated (CLI, Ant, Maven, Gradle, Jenkins DC Plugin Build Step), if the file paths are different at each run, that will certainly cause problems for the Jenkins DC Plugin (Publisher Step) when it tracks changes between builds. If randomly generated temp path is causing problems, perhaps DC core should provide a way to specify this. The Jenkins plugin already provides a way to specify the temp path - this is a global configuration option. If you're using the Jenkins plugin to generate results, have you tried specifying the temp path and if so, do you still get the same results?

**Comment by jeremylong** [comment]
I believe this change was caused by a PR and I missed this in my review. I will go back and review the suspect PR to verify.

**Comment by mikjo** [comment]
@stevespringett — a single run will unpack multiple archives and create multiple unique temporary paths. It's not the `/tmp/dctemp` part that causes noise, it's the `check1868579570325300612tmp/` part that is unique for each artifact being unpacked. (While we are using the jenkins plugin to do the scans in some cases, in cases where we are scanning multiple packages we have switched to command-line invocations, one per package, and then using the jenkins plugin only to summarize the results; so far, the trend data looks more consistent that way because the "related dependency" code seems to be functionally more stable that way. Previously, we got a lot of noise from false changes from instability in the related dependency heuristics. But I don't think that's related to this PR.)

@wschiang told me in conversation that by adding a new test case he believes that he reproduced the problem with 1.3.5 though with a previous test case we hadn't. In practice we went from not seeing this in very large scans (trying to summarize over sets of hundreds or thousands of packages) to seeing many instances, so it seems that if it wasn't uniquely introduced between 1.3.5 and 1.4 it was at least made more likely. (I haven't seen the case of reproducing it with 1.3.5 directly.) The delay in providing a test case is because so far where we reproduced it we had proprietary code in the scan; we were trying to create a non-proprietary, simple, small test case to pinpoint the problem. Our hope was to get a reliable test case and then git bisect to find the offending changeset, but we didn't get there yet.

**Comment by mikjo** [comment]
We now are confident that this is not consistent behavior. We have scanned the same content multiple times and only some of the invocations lose the file path and start tracking `/tmp/dctemp` instead. This makes bisection not particularly relevant for locating the problem, obviously.

It seems to be more likely when we have archives in archives (as a common example, a `lib/` directory in a jar that contains other jars) but we're not sure this is always the case.

Our current workaround is to unpack archive files in the working directory recursively before invoking OWASP DC, where that option is available.

**Comment by mikjo** [comment]
The workaround is inconsistent, and we just saw this happen again with 1.4.3 today

**Comment by jeremylong** [comment]
This is done using the CLI or one of the plugins?

**Comment by mikjo** [comment]
At least in this instance, we are seeing it with the CLI. We are scanning hundreds or thousands of packages (depending on the scan job; some scan larger contexts than others) under Jenkins jobs with one invocation of the CLI per package, and then we do some custom roll up and reporting to aggregate hundreds or thousands of XML report files, as well as letting the Jenkins plug-in aggregate them for viewing from the Jenkins UI. But the Jenkins plug-in is not running dependency-check.

I believe we have also seen it from running the Jenkins plug-in, but that wasn't in use here except to aggregate the resulting XML report files into the web view.

**Comment by mikjo** [comment]
We have finally found a jar in maven central that alone reproduces this issue occasionally though inconsistently in our testing: https://mvnrepository.com/artifact/org.jacoco/org.jacoco.agent/0.7.1.201405082137

We put org.jacoco.agent-0.7.1.201405082137.jar as downloaded from maven central in a directory, and started a loop running the cli scanner against that directory. It is reproducing this somewhat often; we saw it reproduce twice in twelve invocations of the command-line client. We did this test with HotSpot:

```
$ java -version
java version "1.8.0_25"
Java(TM) SE Runtime Environment (build 1.8.0_25-b17)
Java HotSpot(TM) 64-Bit Server VM (build 25.25-b02, mixed mode)
```

I doubt it matters, but we have it configured to do a noupdate scan against a mysql database rather than starting each scan by updating an h2 database.

**Comment by jeremylong** [comment]
@mikjo I truly appreciate the work you've done in isolating an example of this problem. After my talk at JavaOne next week ([depending on vulnerable libraries](https://oracle.rainfocus.com/scripts/catalog/oow16.jsp?event=javaone&search=con6265&search.event=javaone)) I will take a closer look at this and hopefully solve this problem for you.

--Jeremy

**Comment by mikjo** [comment]
You're very welcome, and thank you for raising consciousness of this class of problem.

Our workaround of recursively unpacking in front of the scanning process may be faster than—and at least is not slower than—dynamically unpacking, so this is hitting us less often in our jenkins scans (when it does, we find another archive type to unpack or another job to which to add recursive unpacking). The main impact now is use under the gradle plugin. It was in fact use under the gradle plugin that led to finding the jacoco example. We do not (yet?) have sufficiently-complete false positive suppression across our entire ecosystem to automatically fail developer builds based on OWASP DC failures, so at this point it is noise in the report that does not stop development.

**Comment by mikjo** [comment]
I was wrong about it not stopping development. The exception causes the plugin to fail which stops the build, and I found out today that this is affecting development substantially. I see it happening under gradle both from embedded jars and from pom.xml in jars.

**Comment by jeremylong** [comment]
@mikjo this thread/issue has been about incorrect paths causing duplication of issues, etc. If you are seeing errors that are actually causing the build to fail there are three things that should be done.

1) Ensure that you upgrade to 1.4.3. This resolved some bugs that caused DC to fail.
2) If you are still seeing an issue that causes the build to fail - please open a separate issue
3) While the bug that is causing DC to fail is being fixed you can configure DC with by setting `failOnError=false`.

**Comment by mikjo** [comment]
We're definitely using 1.4.3. I had thought @wschiang had mentioned here that we had seen "error occurred while analyzing '/tmp/dctemp...'" messages associated with the presence of `/tmp/dctemp` in the dependency-check-report.xml files, but I see that I was mistaken. My apologies.

It turns out that the "reproduced" failure with jacoco is of this different kind (or at least different symptom if they do for any reason share a cause) that I will file as a separate bug report. I have just confirmed that the command line is writing dependency-check-report.xml files without `/tmp/dctemp` in them when failing as we see with jacoco on the command line. @wschiang and I miscommunicated about this while working on reporting it. Sorry about that.
